File: internal/server/trend.go
package server

import (
	"fmt"
	"math"
	"slices"
	"strings"

	"github.com/gocov/gocov/internal/store"
)

// The trend chart plots total coverage over the branch's recent uploads
// as server-rendered inline SVG. Geometry lives here so the template only
// places precomputed coordinates.
const (
	trendReportLimit = 60

	trendW, trendH       = 800, 160
	trendPadL, trendPadR = 46, 14
	trendPadT, trendPadB = 12, 24
)

// trendPoint is one upload's marker on the chart.
type trendPoint struct {
	X, Y       float64
	ID         int64
	GateFailed bool
	Title      string // hover text: "2026-08-05 · 82.3% · abc123def456"
}

// trendGridLine is a labeled horizontal gridline (the series min and max).
type trendGridLine struct {
	Y     float64
	Label string
}

// trendThreshold marks the gate minimum on the chart, so the reader sees how
// the series sits against the line that decides pass/fail.
type trendThreshold struct {
	Y     float64
	Label string
}

// trendView is the fully computed chart handed to the repo template.
type trendView struct {
	W, H       int
	X0, X1     int // plot's left/right edge; gridlines and axis labels hang off these
	Branch     string
	Path       string // the polyline through every point
	Points     []trendPoint
	Grid       []trendGridLine
	Thresh     *trendThreshold // gate minimum line; nil when no gate is set
	CurLabel   string          // current (latest) percentage, drawn at the last point
	CurX, CurY float64
	FirstDate  string
	LastDate   string
}

// round1 keeps SVG coordinates at one decimal so output stays compact and
// deterministic.
func round1(v float64) float64 { return math.Round(v*10) / 10 }

// linePath accumulates an SVG polyline — "M x y L x y …" — one point at a
// time, so the chart and the dashboard's sparklines draw a series the same
// way.
type linePath struct{ strings.Builder }

func (p *linePath) lineTo(x, y float64) {
	cmd := " L"
	if p.Len() == 0 {
		cmd = "M"
	}
	fmt.Fprintf(p, "%s%g %g", cmd, x, y)
}

// newTrendView builds the chart for a branch from its merged commit
// reports, given newest-first as ListBranchCommitReports returns them. PR
// reports are excluded: the branch trend reflects the branch's own commits.
// Each report carries the upload id it links to (UploadID). Returns nil when
// fewer than two points remain — the page then omits the section. The gate
// minimum, when the repo has one, is drawn as a dashed threshold line and
// folded into the plotted range so it is always visible.
func newTrendView(branch string, reports []*store.CommitReport, gateMin *float64) *trendView {
	var series []*store.CommitReport // chronological
	for _, report := range slices.Backward(reports) {
		if report.PRID == "" {
			series = append(series, report)
		}
	}
	if len(series) < 2 {
		return nil
	}

	// Y auto-scales to the data's range with padding, not 0-100: a 72-78%
	// story must not flatline. Clamped so the padding never invents
	// impossible percentages.
	lo, hi := series[0].TotalPct, series[0].TotalPct
	for _, u := range series[1:] {
		lo, hi = min(lo, u.TotalPct), max(hi, u.TotalPct)
	}
	pad := max((hi-lo)*0.1, 0.5)
	yLo := max(0, lo-pad)
	yHi := min(100, hi+pad)

	// A configured gate minimum widens the plotted range so its line always
	// lands on the canvas — the grid labels still read the series min/max, so
	// the gate line reads as a separate reference, not a data point.
	if gateMin != nil {
		yLo = min(yLo, max(0, *gateMin))
		yHi = max(yHi, min(100, *gateMin))
	}

	plotW := float64(trendW - trendPadL - trendPadR)
	plotH := float64(trendH - trendPadT - trendPadB)
	x := func(i int) float64 {
		return round1(trendPadL + plotW*float64(i)/float64(len(series)-1))
	}
	y := func(pct float64) float64 {
		return round1(trendPadT + plotH*(yHi-pct)/(yHi-yLo))
	}

	v := &trendView{
		W:         trendW,
		H:         trendH,
		X0:        trendPadL,
		X1:        trendW - trendPadR,
		Branch:    branch,
		FirstDate: series[0].CreatedAt.Format("2006-01-02"),
		LastDate:  series[len(series)-1].CreatedAt.Format("2006-01-02"),
	}
	var path linePath
	for i, u := range series {
		px, py := x(i), y(u.TotalPct)
		path.lineTo(px, py)
		v.Points = append(v.Points, trendPoint{
			X: px, Y: py, ID: u.UploadID, GateFailed: u.GateFailed,
			Title: fmt.Sprintf("%s · %.1f%% · %s", u.CreatedAt.Format("2006-01-02"), u.TotalPct, shortSHA(u.CommitSHA)),
		})
	}
	v.Path = path.String()

	v.Grid = []trendGridLine{{Y: y(hi), Label: fmt.Sprintf("%.1f%%", hi)}}
	if lo != hi {
		v.Grid = append(v.Grid, trendGridLine{Y: y(lo), Label: fmt.Sprintf("%.1f%%", lo)})
	}

	if gateMin != nil {
		v.Thresh = &trendThreshold{Y: y(*gateMin), Label: fmt.Sprintf("gate %.4g%%", *gateMin)}
	}

	last := v.Points[len(v.Points)-1]
	cur := series[len(series)-1]
	v.CurLabel = fmt.Sprintf("%.1f%%", cur.TotalPct)
	v.CurX = last.X
	v.CurY = last.Y - 9
	if v.CurY < 11 { // too close to the top edge: drop below the point
		v.CurY = last.Y + 18
	}
	return v
}

File: web/src/lib/trend.ts
// The coverage chart's geometry, ported from newTrendView in
// internal/server/trend.go so the SPA plots the same shape the server-
// rendered page did. Pure: the component only places these coordinates.

import type { TrendPoint } from "./api/types";
import { shortSha } from "./format";

/** How far left of the last point the current-value label can reach, in viewBox units. */
const LABEL_REACH = 64;
const W = 800;
const H = 160;
const PAD_L = 46;
const PAD_R = 14;
const PAD_T = 12;
const PAD_B = 24;

/** One upload's marker: where it sits, where it links, and what it reads as. */
export interface TrendMark {
  x: number;
  y: number;
  uploadId: number;
  gateFailed: boolean;
  /** "2026-08-05 · 82.3% · abc123def456" */
  label: string;
}

/** A labelled horizontal line: the series min and max, or the gate minimum. */
export interface TrendLine {
  y: number;
  label: string;
}

export interface TrendGeometry {
  w: number;
  h: number;
  /** The plot's left and right edge: lines and axis labels hang off these. */
  x0: number;
  x1: number;
  /** The polyline through every point. */
  path: string;
  marks: TrendMark[];
  grid: TrendLine[];
  /** The gate minimum, drawn dashed; null when the repo has no minimum. */
  threshold: TrendLine | null;
  /** The latest value, labelled at the last point. */
  current: { x: number; y: number; label: string };
  firstDate: string;
  lastDate: string;
}

/** SVG coordinates stay at one decimal, so the markup stays compact. */
const round1 = (v: number) => Math.round(v * 10) / 10;

const day = (iso: string) => iso.slice(0, 10);

const pct1 = (v: number) => v.toFixed(1) + "%";

/** Go's %.4g: four significant digits with trailing zeros dropped. */
const sig4 = (v: number) => String(Number(v.toPrecision(4)));

/**
 * The chart for a branch, given its points oldest first (the API already
 * drops PR reports). Fewer than two points returns null — the page then omits
 * the section. A gate minimum widens the plotted range so its line always
 * lands on the canvas, while the grid labels keep reading the series min and
 * max, so the gate reads as a reference and not as a data point.
 */
export function trendGeometry(points: TrendPoint[], minCoverage: number | null): TrendGeometry | null {
  const first = points[0];
  const last = points[points.length - 1];
  if (points.length < 2 || first === undefined || last === undefined) return null;

  // Y auto-scales to the data's range with padding, not 0–100: a 72–78%
  // story must not flatline. Clamped, so the padding never invents an
  // impossible percentage.
  const values = points.map((p) => p.coverage);
  const lo = Math.min(...values);
  const hi = Math.max(...values);
  const pad = Math.max((hi - lo) * 0.1, 0.5);
  let yLo = Math.max(0, lo - pad);
  let yHi = Math.min(100, hi + pad);
  if (minCoverage !== null) {
    yLo = Math.min(yLo, Math.max(0, minCoverage));
    yHi = Math.max(yHi, Math.min(100, minCoverage));
  }
  const span = yHi - yLo || 1;

  const plotW = W - PAD_L - PAD_R;
  const plotH = H - PAD_T - PAD_B;
  const x = (i: number) => round1(PAD_L + (plotW * i) / (points.length - 1));
  const y = (value: number) => round1(PAD_T + (plotH * (yHi - value)) / span);

  const marks: TrendMark[] = points.map((p, i) => ({
    x: x(i),
    y: y(p.coverage),
    uploadId: p.upload_id,
    gateFailed: p.gate_failed,
    label: `${day(p.at)} · ${pct1(p.coverage)} · ${shortSha(p.sha)}`,
  }));
  const path = marks.map((m, i) => `${i === 0 ? "M" : " L"}${m.x} ${m.y}`).join("");

  const grid: TrendLine[] = [{ y: y(hi), label: pct1(hi) }];
  if (lo !== hi) grid.push({ y: y(lo), label: pct1(lo) });

  const lastMark = marks[marks.length - 1] as TrendMark;
  // The label is right-aligned at the last point, so it stretches back over
  // the points just before it: clear the highest of those, not only the last
  // one — a peak on the second-to-last upload used to sit under the text.
  // When that would leave the canvas, it goes below the lowest of them.
  const under = marks.filter((m) => m.x >= lastMark.x - LABEL_REACH);
  const top = Math.min(...under.map((m) => m.y));
  const labelY = top - 9 < 11 ? Math.max(...under.map((m) => m.y)) + 18 : top - 9;

  return {
    w: W,
    h: H,
    x0: PAD_L,
    x1: W - PAD_R,
    path,
    marks,
    grid,
    threshold: minCoverage === null ? null : { y: y(minCoverage), label: `gate ${sig4(minCoverage)}%` },
    current: { x: lastMark.x, y: labelY, label: pct1(last.coverage) },
    firstDate: day(first.at),
    lastDate: day(last.at),
  };
}
